### DISCUSSION TITLE
How can a global dispatcher be used in tests while a local dispatcher is used for proxy support?

### DISCUSSION THREAD

**Discussion by parkerbxyz** [nodejs/undici#3792] [Q&A]
We are using a local `EnvHttpProxyAgent` dispatcher for proxy support. For testing, we have a global `MockAgent` dispatcher set. The global dispatcher is not being used when the local dispatcher is set, as demonstrated in the following example:

```js
import { EnvHttpProxyAgent, MockAgent, setGlobalDispatcher } from "undici";

const mockAgent = new MockAgent();
mockAgent.disableNetConnect();
setGlobalDispatcher(mockAgent);
const mockPool = mockAgent.get("https://example.test");

mockPool
  .intercept({
    path: `/`,
    method: "GET",
  })
  .reply(200);

await fetch("https://example.test", {
  // Comment out the following line to see the test succeed
  dispatcher: new EnvHttpProxyAgent(),
});

console.log("OK");
```

Is this the intended behavior, or is this a bug? If it is the intended behavior, is there another way we can use a global dispatcher in tests without it being overridden by the local dispatcher?

**Comment by parkerbxyz** [comment]
Possibly related: https://github.com/nodejs/undici/issues/3444

**Reply by metcoder95** [reply]
Just to clarify, are you looking to also mock the `ProxyAgent`, right?

**Reply by parkerbxyz** [reply]
We are not currently, but that is the plan, yes.
